In [173]:

# Imports
from math import sqrt
from scipy import stats
from pydataset import data
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# Attendance Data

In [18]:
# lets look at this data:
df = pd.read_csv('untidy-data/attendance.csv')
df.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [19]:
df.shape

(4, 9)

In [20]:
df.columns

Index(['Unnamed: 0', '2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04',
       '2018-01-05', '2018-01-06', '2018-01-07', '2018-01-08'],
      dtype='object')

In [43]:
# Melt the data
#id_vars are the columns you want to keep, NOT melt
df_melt = pd.melt(df, id_vars= 'Unnamed: 0', var_name='date', value_name='attendance')
df_melt.head()

Unnamed: 0        date attendance
0      Sally  2018-01-01          P
1       Jane  2018-01-01          A
2      Billy  2018-01-01          A
3       John  2018-01-01          P
4      Sally  2018-01-02          T

In [44]:
#renaming the column name from Unnamed: 0 to name
df_melt = df_melt.rename(columns={'Unnamed: 0':'name'})
df_melt.head()

name        date attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

In [46]:
df_melt['attendance'] = df_melt['attendance'].replace({'T':0.9, 'P':1, 'A':0.0, 'H':0.5})

In [47]:
df_melt.head()

name        date  attendance
0  Sally  2018-01-01         1.0
1   Jane  2018-01-01         0.0
2  Billy  2018-01-01         0.0
3   John  2018-01-01         1.0
4  Sally  2018-01-02         0.9

In [52]:
df_melt.groupby(by=['name']).mean()

attendance
name             
Billy      0.5250
Jane       0.6875
John       0.9125
Sally      0.7625

# Coffee Levels

- a. Read the coffee_levels.csv file.

In [57]:
df = pd.read_csv('untidy-data/coffee_levels.csv')
df.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

- b. Transform the data so that each carafe is in it's own column.

In [61]:
#reset_index() = makes the data flat
df.shape

(30, 3)

In [63]:
df.columns

Index(['hour', 'coffee_carafe', 'coffee_amount'], dtype='object')

In [66]:
#pivot the data
df_pivot = df.pivot(index='hour', columns = 'coffee_carafe', values='coffee_amount') 
df_pivot

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
9              0.451018  0.521502  0.915990
10             0.843279  0.023163  0.144928
11             0.335533  0.235529  0.311495
12             0.898291  0.017009  0.771947
13             0.310711  0.997464  0.398520
14             0.507288  0.058361  0.864464
15             0.215043  0.144644  0.436364
16             0.183891  0.544676  0.280621
17             0.391560  0.594126  0.436677

- c. Is this the best shape for the data?
     - Depending on what this data is being used for, it is not the most user friendly viewing of the data. 

# Cake Recipes

- a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [103]:
df = pd.read_csv('untidy-data/cake_recipes.csv')

In [104]:
df

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

- b. Tidy the data as necessary. 

In [105]:
df_tidy = pd.melt(df, id_vars='recipe:position', var_name='temp', value_name='taste_score')
df_tidy.head(5)

recipe:position temp  taste_score
0        a:bottom  225    61.738655
1           a:top  225    51.709751
2        b:bottom  225    57.095320
3           b:top  225    82.455004
4        c:bottom  225    96.470207

In [106]:
#rename column to split / expand:
df_tidy = df_tidy.rename(columns={'recipe:position':'recipe_position'})
df_tidy.head()

recipe_position temp  taste_score
0        a:bottom  225    61.738655
1           a:top  225    51.709751
2        b:bottom  225    57.095320
3           b:top  225    82.455004
4        c:bottom  225    96.470207

In [107]:
df_tidy1 = df_tidy.recipe_position.str.split(':', expand=True)
df_tidy1.columns = ['recipe', 'position']
df_tidy1.head()

recipe position
0      a   bottom
1      a      top
2      b   bottom
3      b      top
4      c   bottom

In [114]:
df_tidy2 = pd.concat([df_tidy1, df_tidy], axis=1).drop(columns='recipe_position')
df_tidy2.head()

recipe position temp  taste_score
0      a   bottom  225    61.738655
1      a      top  225    51.709751
2      b   bottom  225    57.095320
3      b      top  225    82.455004
4      c   bottom  225    96.470207

- c. Which recipe, on average, is the best? 
    - recipe b

In [186]:
best_recipe = df_tidy2.groupby('recipe').taste_score.mean()
best_recipe.nlargest(1)

recipe
b    76.736074
Name: taste_score, dtype: float64

- d. Which oven temperature, on average, produces the best results? 
     - 275

In [184]:
best_temp = df_tidy2.groupby('temp').taste_score.mean()
best_temp.nlargest(1)

temp
275    74.886754
Name: taste_score, dtype: float64

- e. Which combination of recipe, rack position, and temperature gives the best result?
  - recipe b, bottom rack, 300 degrees

In [178]:
df_tidy2.nlargest(1, ['taste_score'])

recipe position temp  taste_score
26      b   bottom  300    99.248541